### PR TITLE
[TACHYON-255] Support for OpenStack Swift

### PR DESCRIPTION
**Pull request by gilv**
Support for OpenStack Swift and SoftLayer public object store
What is not here:
*********************
1. Documentation. Will submit it later in separate patch.
2. Testing is not here yet. Will be added later

How it works:
***************
First of all you need some local Swift or an account account in SoftLayer.
Those who like to setup local Swift - it's a very easy procedure ( see http://docs.openstack.org/developer/swift/development_saio.html )

My patch uses hadoop-openstack project, which internally implements Swift API and expose HDFS interface, so hadoop-openstack can easily be plugged into Tachyon.
It's a stable project, part of Hadoop releases from 2.4.0.  Hadoop community provides fixes and improves this driver constantly.

Few known issues:
***********************
Swift may come with various authentication models. 
It shipped with v1 authentication model, based on temp auth. However many uses v2 authentication model, based on Keystone. Current release of hadoop-openstack project has implementation based on Keystone. This means that your local Swift should be configured with Keystone.
SoftLayer doesn't use Keystone and use v1 auth model instead. The patch for the hadoop-driver to support v1 is not yet merged into Hadoop. 
Therefore those who wish to use SoftLayer object store need to follow more instructions ( until the pending patch will be merged into Hadoop)
The instructions are here https://gist.github.com/gilv/9d1e60aef201e76ffdba ( just follow Hadoop part section )

How to configure Tachyon
********************************
1. /conf folder contains core-site.xml.template
This file contains three sample configurations, one for Dallas, second local Swift with Keystone, third local Swift with temp auth.Depends on what you use, you need to configure them and add credentials and access URLs.

2. Rename core-site.xml.template file to core-site.xml and keep it in the same directory ( or not...it's configurable option in tachyon-end.sh or in 
tachyon.properties under tachyon.underfs.hadoop.configuration=${tachyon.home}/conf/core-site.xml

3. See tachyon-env.sh.swift how to run it. There is basically only one modification.

I tested that ./bin/tachyon runTest Basic CACHE_THROUGH works and also managed to sync existing objects in Swift with Tachyon.

### CODE REVIEW

**Comment by haoyuan** [comment]
Jenkins, test this please.

File: pom.xml [hunk 1]
```diff
@@ -124,7 +124,7 @@
     <test.output.redirect>true</test.output.redirect>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <java.version>1.6</java.version>
-    <hadoop.version>1.0.4</hadoop.version>
+    <hadoop.version>2.6.0</hadoop.version>
```
Comment (rvesse) [inline pom.xml]: While I personally would love Tachyon to target Hadoop 2.x by default I expect that you will get some pushback on this particular change

File: underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java [hunk 2]
```diff
@@ -66,9 +66,10 @@ public HdfsUnderFileSystem(String fsDefaultName, TachyonConf tachyonConf, Object
     } else {
       tConf = new Configuration();
     }
-
     prepareConfiguration(fsDefaultName, tachyonConf, tConf);
-
+    Utils.addKey(tConf, tachyonConf, "tachyon.underfs.hadoop.configuration");
```
Comment (rvesse) [inline underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java]: Is there any reason for this not to go in the `prepareConfiguration()` method?

File: pom.xml (hunk 1 above)
Comment (gilv) [inline pom.xml]: Ohh...good point. I will remove this one.

File: underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java (hunk 2 above)
Comment (gilv) [inline underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystem.java]: I will move this line to prepareConfiguration() as you suggested.

**Comment by calvinjia** [comment]
@gilv Thanks for the patch! Could you include the documentation in docs/Setup-UFS.md?

**Comment by haoyuan** [comment]
Jenkins, add to whitelist.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by gilv** [comment]
i will check why TachyonConfTest failed

**Comment by gilv** [comment]
right, i see why it failed. i did some modifications to the properties, before i change it back - i would like to ask a question related it:

1. So the original value of tachyon.master.journal.folder was pointed to ${tachyon.home}/journal/.
It looks like to me that it's wrong...If someone use particular file system implementation - he would want to store all there and not to store journal in the home directory
2.  The same related tachyon.master.temporary.folder, i changed it instead of /tmp to ${tachyon.underfs.address}/tmp... 

Should i propose this as a separate patch or rollback completely - or you agree with me and i should modify tests?

-tachyon.master.journal.folder=${tachyon.home}/journal/
+#tachyon.master.journal.folder=${tachyon.home}/journal/
+tachyon.master.journal.folder=${tachyon.underfs.address}/journal/
tachyon.master.format.file_prefix=_format_
tachyon.master.port=19998
tachyon.master.web.port=19999
-tachyon.master.temporary.folder=/tmp
+tachyon.master.temporary.folder=${tachyon.underfs.address}/tmp

**Comment by rvesse** [comment]
The journal is only relevant for the master hence why it is placed on the local file system.  Also the journal needs to be as fast to access as possible in order to avoid slowdowns because Tachyon uses the journal as a write ahead log so updates to the file system metadata are gated by the speed of the journal access.

The journal should only be placed on a shared file system if you are setting up fault tolerance per the [Fault Tolerant Tachyon Cluster](http://tachyon-project.org/Fault-Tolerant-Tachyon-Cluster.html) documentation in which case all the masters need to be able to access the journal.

Similarly the master temporary folder is just that - temporary - so again it makes sense for it to be local to the master since it doesn't need to be available anywhere else in the cluster and access to it wants to be as fast as possible

**Comment by calvinjia** [comment]
@gilv Could you update this with the latest master?

**Comment by gilv** [comment]
It failed since Hadoop-openstack is not exists in the Hadoop 1.X.X..
Is there way to modify pom.xml so it will be included only for Hadoop 2.4.0 and up? Or i should move it to comment?

[ERROR] Failed to execute goal on project tachyon-underfs-hdfs: Could not resolve dependencies for project org.tachyonproject:tachyon-underfs-hdfs:jar:0.7.0-SNAPSHOT: Could not find artifact org.apache.hadoop:hadoop-openstack:jar:1.0.4 in central (https://repo1.maven.org/maven2) -> [Help 1]

**Comment by calvinjia** [comment]
@gilv Its my suspicion that lower versions of hadoop (ie. 1.x) will not work with the hadoop openstack client (2.x). Its worth a try to see if it can work though. In case it does, you can just use another variable to specify the hadoop openstack version since its no longer tied to the actual hadoop version. In the case it doesn't work, one solution would be to modularize it like glusterfs and we document that hadoop 2.x is required to use it.

**Comment by gilv** [comment]
@calvinjia Hadoop-openstack requires version 2.4.0 and up, it will not work with versions 1.x.x
I will try to look how you resolved it for GlusterFS.

File: core/pom.xml [hunk 3]
```diff
@@ -96,8 +96,13 @@
     <dependency>
       <groupId>commons-httpclient</groupId>
       <artifactId>commons-httpclient</artifactId>
-      <version>3.0.1</version>
-      <scope>test</scope>
+      <version>3.1</version>
+      <!--scope>test</scope-->
```
Comment (calvinjia) [inline core/pom.xml]: Is this required by hadoop-openstack?

**Comment by calvinjia** [comment]
@gilv Could you modularize the swift integration into another module under `underfs` like `underfs/swift` (similar to `underfs/glusterfs`)? Although there is not too much code change in the hdfs underfilesystem, I think it would be cleaner to keep it separate from hdfs, especially if we add changes in the future.

**Comment by gilv** [comment]
@calvinjia : But the point is to use HDFS interfaces, why would you like to have the same code twice? The hdfs module you have implements HDFS interfaces and than we can use it for Swift or S3 or HDFS... I think it's very good and make code cleaner, otherwise i will need to copy 99% of the code to /swift module...

File: core/pom.xml (hunk 3 above)
Comment (gilv) [inline core/pom.xml]: Yes, 
should i add it also dependent on hadoop-openstack,version?

**Comment by calvinjia** [comment]
You should be able to extend the hdfs interface and override the prepare configuration portion? Or do you mean there's just too much boiler plate to write for a new submodule?

**Comment by calvinjia** [comment]
@gilv I think the change is minor enough to bundle it into the hdfs implementation. In the future if there are more specific things that won't be encapsulated by the conf, we can think about moving to another submodule.

I have two other concerns:

1. I think the open stack dependency will bring in dependencies like hadoop-common and other libraries that Tachyon already has. We should manage these versions because if a user wishes to use hadoop 1.x, openstack 2.x will still be included and possibly bring in conflicting libraries. 

2. Could you add documentation to docs/Setup-UFS.md?

**Comment by gilv** [comment]
@calvinjia : I will add documentation. sure.
Related your concern about dependencies: what is the reason Tachyon doesn't use Hadoop 2.4.0 by default (or up) for hdfs module? In any case it just some dependencies and no  one really need to install Hadoop for this, so why not to addapt hdfs module to use 2.4.0?

**Comment by calvinjia** [comment]
@gilv 
Currently Tachyon's default version is 1.0.4, but I can see an argument for using 2.x. But if the user wishes to use Tachyon with Hadoop 1.x, we should still support that. Bringing in 2.x hadoop dependencies in the hdfs module (from open stack, since its included regardless of if the user is intending to use it) will probably cause issues when the 1.x hadoop client is used.

**Comment by gilv** [comment]
@calvinjia, i also think there will be issues with Hadoop 1.x client. Does GlusterFS has the same issues?  Is there a way to modify pom file so that it will use Hadoop-openstack only when Tachyon compiled with Hadoop 2,4.0 and up?

**Comment by calvinjia** [comment]
@gilv Actually since the code doesn't use anything from hadoop-openstack directly, you could just include it in a profile so by default it would not be included. If the user wants to use swift they would need to rebuild Tachyon like `mvn clean install -Pswift` if the profile was called swift.

**Comment by calvinjia** [comment]
@gilv could you rebase/merge with master again? Seems like there are some conflicts.

File: core/pom.xml [hunk 4]
```diff
@@ -119,6 +119,70 @@
...
+                </configuration>
+                <goals>
+                  <goal>testCompile</goal>
+                </goals>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
```
Comment (calvinjia) [inline core/pom.xml]: You should be able to add a simpler profile to `underfs/hdfs/pom.xml` just including the open-stack dependency.

**Comment by gilv** [comment]
this also doesn't work for me now...something is wrong after latest refactoring, trying to debug it.

**Comment by gilv** [comment]
@calvinjia Do you have any idea why i am getting this? 

org.apache.hadoop.fs.FileSystem: Provider tachyon.hadoop.TFS not found

	at tachyon.underfs.UnderFileSystemRegistry.create(UnderFileSystemRegistry.java:132)
	at tachyon.underfs.UnderFileSystem.get(UnderFileSystem.java:85)
	at tachyon.underfs.UnderFileSystem.get(UnderFileSystem.java:69)
	at tachyon.Format.formatFolder(Format.java:38)
	at tachyon.Format.main(Format.java:66)

**Comment by calvinjia** [comment]
@gilv have you synced to the latest master?

**Comment by gilv** [comment]
@calvinjia the last time i updated was two days ago and got those exceptions and was sure it come from my code... I just updated as you suggested and they are gone. What was that exception? I tried to debug it and couldn't understand the origin of it

**Comment by calvinjia** [comment]
@gilv That exception probably occurred because the hadoop FileSystem meta-inf file was in the wrong place (stating that an implementation, `tachyon.hadoop.TFS` exists in the module but really doesn't). Lots of files were moved around in the past few days and its possible that there was a bad state at some time.

**Comment by gilv** [comment]
@calvinjia : I saw the recent structure of underfs module. I am 100% agree now with your previous suggestion to make a separate module for Swift. I think it's much cleaner and can be better supported for specific Swift needs.

**Comment by calvinjia** [comment]
@gilv 
Thanks for making a new submodule!

Instead of extending `UnderFileSystem`, it may be cleaner to extend `HdfsUnderFileSystem` and override only the methods that require special changes for swift.

Could you also take a look at a underfs submodule pom (ie. `underfs/hdfs/pom.xml`) and see how we are structuring the dependencies after the work of splitting up the core jar?

**Comment by gilv** [comment]
@calvinjia Good points. Please review my changes.

**Comment by calvinjia** [comment]
@gilv Looks good overall. I think you can revert the changes made to `core/pom.xml`?

**Comment by gilv** [comment]
@calvinjia I am not sure why, but without modification to core/pom.xml - it doesn't work.

**Comment by gilv** [comment]
@calvinjia : you mean to remove the swift profile from core/pom.xml? it doesn't work without it, Not sure why

**Comment by calvinjia** [comment]
@gilv I actually mean all the changes done to `core/pom.xml`. What kind of error do you get when it is reverted?

**Comment by gilv** [comment]
@calvinjia  If there are no further remarks, can we merge it?. Changes to core/pom.xml will be in separate patch? Thanks

**Comment by calvinjia** [comment]
@gilv After merging master you should be able to revert the changes in `core/pom.xml` and everything should still work. You will need to add the new swift module to `assembly/pom.xml`.

**Comment by gilv** [comment]
@calvinjia not sure why it happens, but the some of merge was manual and than it also included my changes to core/pom.xml. This is why this commit has all the changes from master...

**Comment by gilv** [comment]
is there anything else? Can we merge it?

**Comment by calvinjia** [comment]
@gilv it seems like the changes to `core/pom.xml` were not reverted? Or did you mean that you couldn't due to some git weirdness?

**Comment by gilv** [comment]
@calvinjia  I did changes to it, i reverted all changes. will check it again now

**Comment by gilv** [comment]
i don't see core module anymore...you make changes to rapidly :)

**Comment by calvinjia** [comment]
@gilv it's renamed to tachyon-servers now

**Comment by gilv** [comment]
@calvinjia  so i don't have any modification to servers/pom.xml, there is no single word about openstack in this file, what changes you mean i  need to revert there?

File: underfs/swift/src/test/java/tachyon/underfs/swift/SwiftUnderFileSystemTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import tachyon.underfs.UnderFileSystemRegistry;
+
+public class SwiftUnderFileSystemTest {
+
+  @Test
+  public void factoryTest() {
+    TachyonConf conf = new TachyonConf();
+    
+    UnderFileSystemFactory factory =  UnderFileSystemRegistry.find("swift://localhost/test/path", conf);
+    Assert.assertNotNull("A UnderFileSystemFactory should exist for Swift paths when using this module", factory);
```
Comment (haoyuan) [inline underfs/swift/src/test/java/tachyon/underfs/swift/SwiftUnderFileSystemTest.java]: much longer than 100 chars. 

why doesn't the style checker check this?

File: underfs/swift/src/main/java/tachyon/underfs/swift/SwiftUnderFileSystemFactory.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  @Override
+  public boolean supportsPath(String path, TachyonConf conf) {
+    if (path == null) {
+      return false;
+    }
+    if (path.startsWith("swift://")) {
+      return true;
+    }
+    return false;
+
```
Comment (haoyuan) [inline underfs/swift/src/main/java/tachyon/underfs/swift/SwiftUnderFileSystemFactory.java]: extra empty line.

File: underfs/swift/src/main/java/tachyon/underfs/swift/SwiftUnderFileSystem.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import tachyon.Constants;
+import tachyon.underfs.hdfs.HdfsUnderFileSystem;
+
+/**
+ * Swift UnderFilesystem implementation
+ */
+public class SwiftUnderFileSystem extends HdfsUnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+
```
Comment (haoyuan) [inline underfs/swift/src/main/java/tachyon/underfs/swift/SwiftUnderFileSystem.java]: extra line.

File: servers/pom.xml [hunk 8]
```diff
@@ -94,13 +94,12 @@
...
     </dependency>
-
     <!-- Test Dependencies -->
     <dependency>
       <groupId>commons-httpclient</groupId>
       <artifactId>commons-httpclient</artifactId>
-      <version>3.0.1</version>
-      <scope>test</scope>
+      <version>3.1</version>
+      <!--scope>test</scope-->
```
Comment (haoyuan) [inline servers/pom.xml]: why was this changed?

File: servers/pom.xml [hunk 9]
```diff
@@ -119,6 +118,51 @@
...
+      <id>contractTest</id>
+      <dependencies>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-minicluster</artifactId>
+        </dependency>
+      </dependencies>
+    </profile>
+
+  </profiles>
```
Comment (calvinjia) [inline servers/pom.xml]: These changes shouldn't be needed.

File: servers/pom.xml (hunk 8 above)
Comment (calvinjia) [inline servers/pom.xml]: Let's change this back to test and then include it in the swift pom.

File: underfs/swift/src/test/resources/test_metrics.properties [hunk 10]
```diff
@@ -0,0 +1,6 @@
+*.sink.console.class=tachyon.metrics.sink.ConsoleSink
+*.sink.console.period=15
+*.source.jvm.class=tachyon.metrics.source.JvmSource
+master.sink.console.period=20
+master.sink.console.unit=minutes
+master.sink.jmx.class=tachyon.metrics.sink.JmxSink
```
Comment (calvinjia) [inline underfs/swift/src/test/resources/test_metrics.properties]: This file is no longer necessary.

**Comment by gilv** [comment]
@haoyuan : I run mvn checkstyle on swift module

[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Tachyon Under File System - Swift 0.7.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-checkstyle-plugin:2.13:checkstyle (default-cli) @ tachyon-underfs-swift ---
[INFO] Starting audit...
Audit done.

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.309s
[INFO] Finished at: Fri May 15 11:21:47 IDT 2015
[INFO] Final Memory: 24M/217M
[INFO] ------------------------------------------------------------------------

**Comment by calvinjia** [comment]
A few minor formatting issues, but overall LGTM! Thanks for doing this PR despite the refactoring of the overall codebase, its a very important feature.

**Comment by haoyuan** [comment]
Thanks for the great feature!
